### PR TITLE
Support locked entities, and headless video recording using sim time

### PR DESCRIPTION
**Pull request by nkoenig**
Signed-off-by: Nate Koenig <nate@openrobotics.org>

# 🎉 New feature

Related PR: https://github.com/ignitionrobotics/ign-sensors/pull/137

## Summary

The goal of this PR is to support generating videos from log files using a headless simulation instance. In order to accomplish this, I need to spawn a camera into simulation and keep that camera around during log seeks. This lead to the need for `locked` entities within the ECM.

Recording a video using simulation time is required when you want a video to playback at real-time speeds. I copied the sim time video recording code from the Scene3d plugin. 

## Test it

There are ECM tests for the locking feature.
The easiest way to test the sim time video recording is to use the  `camera_video_record_dbl_pendulum.sdf` example world file, and add `<use_sim_time>true</use_sim_time>` to the plugin. Depending on your machine, you may have to adjust the physics step to make sure simulation is running slower than real-time. Record a video, and note the simulation time duration for the video. The resulting video should have matching duration, which can be inspected using:

`mplayer -identify VIDEO.MP4 2>/dev/null | grep ID_LENGTH`

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: include/ignition/gazebo/EntityComponentManager.hh [hunk 1]
```diff
@@ -91,6 +91,35 @@ namespace ignition
       public: void RequestRemoveEntity(const Entity _entity,
           bool _recursive = true);
 
+      /// \brief Prevent an entity and optionally its children from
+      /// being removed.
```
Comment (chapulina) [inline include/ignition/gazebo/EntityComponentManager.hh:95]: The word "lock" implies to me that more is being prevented than just removal. Without reading these docs, I assumed the function would prevent the entity's components from being modified.

How about something like `MarkAsUnremovable`?

**Comment by iche033** [comment]
The sim time video recording feature is working for me. I also tried running faster than real time and that's working as well and produced the correct video length. (RTF just drops a little when video recording is on). 

another nice video recorder param that we could port over later from Scene3D is <bitrate> so users can record high quality videos.

**Comment by iche033** [comment]
> another nice video recorder param that we could port over later from Scene3D is so users can record high quality videos.

#864

File: include/ignition/gazebo/EntityComponentManager.hh (hunk 1 above)
Comment (nkoenig) [reply include/ignition/gazebo/EntityComponentManager.hh:95]: Changed the name in 1e908d310e46dfd94c1529c1a658ac8ec09397df.

**Review by mjcarroll — APPROVED** [review]
LGTM with Green CI (minus tests we know are flaky)

**Review by mjcarroll — APPROVED** [review]

**Pull request merged**